File: leetcode/C++/minimize-deviation-in-array.cpp
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        int mn=1000000001, sol=1000000001, mx=0;
        priority_queue<int> pq;
        for(int &n : nums){
            if(n&1) n*=2;
            mn=min(mn, n);
            pq.push(n);
        }
        while(!pq.empty()){
            mx=pq.top(); pq.pop(); sol=min(sol, mx-mn); 
            if(mx&1) break;
            mx/=2; mn=min(mn, mx); pq.push(mx);
        }
        return sol;
    }
};

File: leetcode/Python/minimize-deviation-in-array.py
class Solution:
    def minimumDeviation(self, nums: List[int]) -> int:
        sol, mx, mn, pq = 1000000001, 0, 1000000001, list()
        for i in range(len(nums)):
            if nums[i]&1 > 0:
                nums[i]*=2;
            mn = min(mn, nums[i])
            heapq.heappush(pq, (-nums[i], nums[i]))
        while len(pq) > 0:
            _, mx = heapq.heappop(pq) 
            sol = min(sol, mx-mn)
            if mx&1 > 0:
                break
            mx//=2; mn=min(mn, mx); heapq.heappush(pq, (-mx, mx))
        return sol